File: 0.Alcor/Gobang.m
clc, clear all, close all;


setGobang = zeros(15);
setlist = [];


clc;
plotGobang(setGobang)

% Alcor-AI

% AlcorBoW = 2; % b = 1, w = 2;
% for kk = 1 : 225
%     
%     if mod(kk + AlcorBoW - 1, 2) == 0
%         [setGobang, scoreGobang, ii, jj] = naiveAI(setGobang, mod(AlcorBoW + 1, 2));
%         setlist = [setlist; [ii, jj]];
%     else
%         if AlcorBoW == 1
%                 X_bset = double(setGobang == 1);
%                 X_wset = double(setGobang == -1);
%                 X_eset = double(setGobang == 0);
%         elseif AlcorBoW == 2
%                 X_bset = double(setGobang == -1);
%                 X_wset = double(setGobang == 1);
%                 X_eset = double(setGobang == 0);
%         end
%         X_turn1 = zeros(15);
%         X_turn2 = zeros(15);
%         X_turn3 = zeros(15);
%         X_turn4 = zeros(15);
%         X_turn5 = zeros(15);
%         for kk = 1 : size(setlist, 1)
%             if kk == 5
%                 X_turn5 = setGobang;
%                 X_turn5(setGobang == -1) = 1;
%                 X_turn5(setlist(end, 1), setlist(end, 2)) = 0;
%                 X_turn5(setlist(end - 1, 1), setlist(end - 1, 2)) = 0;
%                 X_turn5(setlist(end - 2, 1), setlist(end - 2, 2)) = 0;
%                 X_turn5(setlist(end - 3, 1), setlist(end - 3, 2)) = 0;
%                 break;
%             elseif kk == 1
%                 X_turn1(setlist(end, 1), setlist(end, 2)) = 1;
%             elseif kk == 2
%                 X_turn2(setlist(end - 1, 1), setlist(end - 1, 2)) = 1;
%             elseif kk == 3
%                 X_turn3(setlist(end - 2, 1), setlist(end - 2, 2)) = 1;
%             elseif kk == 4
%                 X_turn4(setlist(end - 3, 1), setlist(end - 3, 2)) = 1;
%             end
%         end
%         X(:, :, 1) = X_bset;
%         X(:, :, 2) = X_wset;
%         X(:, :, 3) = X_eset;
%         X(:, :, 4) = X_turn1;
%         X(:, :, 5) = X_turn2;
%         X(:, :, 6) = X_turn3;
%         X(:, :, 7) = X_turn4;
%         X(:, :, 8) = X_turn5;
%         
%         [ii, jj] = Alcor(X);
%         
%         
%         if setGobang(ii, jj) == 0
%             setGobang(ii, jj) = (-1) ^ mod(AlcorBoW + 1, 2);
%         else
%             fprintf('Error!!!!\n');
%         end
%         setlist = [setlist; [ii, jj]];
%     end
%     clc;
%     plotGobang(setGobang)
%     fprintf('%d, %d\n', ii, jj)
%     gg = judge(setGobang, ii, jj);
%     if gg == 1
%         break;
%     end
%     pause;
% end



% Human-AI

AlcorBoW = 2; % b = 1, w = 2;
for kk = 1 : 225
    
    if mod(kk + AlcorBoW - 1, 2) ~= 0
        if AlcorBoW == 1
                X_bset = double(setGobang == 1);
                X_wset = double(setGobang == -1);
                X_eset = double(setGobang == 0);
        elseif AlcorBoW == 2
                X_bset = double(setGobang == -1);
                X_wset = double(setGobang == 1);
                X_eset = double(setGobang == 0);
        end
        X_turn1 = zeros(15);
        X_turn2 = zeros(15);
        X_turn3 = zeros(15);
        X_turn4 = zeros(15);
        X_turn5 = zeros(15);
        for kk = 1 : size(setlist, 1)
            if kk == 5
                X_turn5 = setGobang;
                X_turn5(setGobang == -1) = 1;
                X_turn5(setlist(end, 1), setlist(end, 2)) = 0;
                X_turn5(setlist(end - 1, 1), setlist(end - 1, 2)) = 0;
                X_turn5(setlist(end - 2, 1), setlist(end - 2, 2)) = 0;
                X_turn5(setlist(end - 3, 1), setlist(end - 3, 2)) = 0;
                break;
            elseif kk == 1
                X_turn1(setlist(end, 1), setlist(end, 2)) = 1;
            elseif kk == 2
                X_turn2(setlist(end - 1, 1), setlist(end - 1, 2)) = 1;
            elseif kk == 3
                X_turn3(setlist(end - 2, 1), setlist(end - 2, 2)) = 1;
            elseif kk == 4
                X_turn4(setlist(end - 3, 1), setlist(end - 3, 2)) = 1;
            end
        end
        X(:, :, 1) = X_bset;
        X(:, :, 2) = X_wset;
        X(:, :, 3) = X_eset;
        X(:, :, 4) = X_turn1;
        X(:, :, 5) = X_turn2;
        X(:, :, 6) = X_turn3;
        X(:, :, 7) = X_turn4;
        X(:, :, 8) = X_turn5;
        
        [ii, jj] = Alcor(X);
        
        
        if setGobang(ii, jj) == 0
            setGobang(ii, jj) = (-1) ^ mod(AlcorBoW + 1, 2);
        else
            fprintf('Error!!!!\n');
        end
        setlist = [setlist; [ii, jj]];
    else
        ii = input('column = ');
        jj = input('row = ');
        if setGobang(ii, jj) == 0
            setGobang(ii, jj) = (-1) ^ mod(AlcorBoW, 2);
        end
    end
    clc;
    plotGobang(setGobang)
    fprintf('%d, %d\n', ii, jj)
    gg = judge(setGobang, ii, jj);
    if gg == 1
        break;
    end
end

File: Gobang_python/Gobang.py
# Initialization
from __future__ import print_function
from numpy import *
import numpy as np
import copy
import math
import support
import support2
import pickle

AlcorBoW = 2;

setGobang = np.zeros([15, 15]);
setlist = [];
if AlcorBoW == 1:
    Theta = open('stackedAEThetaBlack.pkl', 'rb')
    #Theta = open('/storage/emulated/0/qpython/work/Gobang/stackedAEThetaBlack.pkl', 'rb')
    pass
elif AlcorBoW == 2:
    Theta = open('stackedAEThetaWhite.pkl', 'rb')
    #Theta = open('/storage/emulated/0/qpython/work/Gobang/stackedAEThetaWhite.pkl', 'rb')
    pass
stackedAETheta = pickle.load(Theta);
netconfig = [1800, 500, 500, 300];
outputSize = 225;


for kk in range(225):
    if (kk + AlcorBoW) % 2 == 0:
        setGobang, scoreGobang, ii, jj = support.naiveAI(setGobang, mod(AlcorBoW + 1, 2));
        setlist.extend([0]);
        setlist[kk] = [[ii], [jj]];
        pass
    else:
        if AlcorBoW == 1:
            X_bset = double(setGobang == 1);
            X_wset = double(setGobang == -1);
            X_eset = double(setGobang == 0);
            pass
        elif AlcorBoW == 2:
            X_bset = double(setGobang == -1);
            X_wset = double(setGobang == 1);
            X_eset = double(setGobang == 0);
            pass
        X_turn1 = np.zeros([15, 15]);
        X_turn2 = np.zeros([15, 15]);
        X_turn3 = np.zeros([15, 15]);
        X_turn4 = np.zeros([15, 15]);
        X_turn5 = np.zeros([15, 15]);
        for kk2 in range(len(setlist)):
            if kk2 == 4:
                X_turn5 = copy.deepcopy(setGobang);
                X_turn5[setGobang == -1] = 1;
                X_turn5[setlist[-1][0], setlist[-1][1]] = 0;
                X_turn5[setlist[-2][0], setlist[-2][1]] = 0;
                X_turn5[setlist[-3][0], setlist[-3][1]] = 0;
                X_turn5[setlist[-4][0], setlist[-4][1]] = 0;
                break
                pass
            elif kk2 == 0:
                X_turn1[setlist[-1][0], setlist[-1][1]] = 1;
                pass
            elif kk2 == 1:
                X_turn2[setlist[-2][0], setlist[-2][1]] = 1;
                pass
            elif kk2 == 2:
                X_turn3[setlist[-3][0], setlist[-3][1]] = 1;
                pass
            elif kk2 == 3:
                X_turn4[setlist[-4][0], setlist[-4][1]] = 1;
                pass
            pass
        X = [];
        X.extend([X_bset]);
        X.extend([X_wset]);
        X.extend([X_eset]);
        X.extend([X_turn1]);
        X.extend([X_turn2]);
        X.extend([X_turn3]);
        X.extend([X_turn4]);
        X.extend([X_turn5]);
        ii, jj = support2.Alcor(X, stackedAETheta, netconfig, outputSize);
        if setGobang[ii, jj] == 0:
            setGobang[ii, jj] = pow(-1, (AlcorBoW + 1) % 2);
            pass
        else:
            print("ERROR!!!")
            pass
        setlist.extend([0]);
        setlist[kk] = [[ii], [jj]];
        pass
    support.plotGobang(setGobang);
    gg = support.judge(setGobang, ii, jj);
    if gg == 1:
        print('Black Win!!')
        break
    elif gg == -1:
        print('White Win!!')
        break
    pass


